src/app/pcasts/dao/listening_histories_dao.py

```diff
@@ -1,25 +1,46 @@
 from app.pcasts.dao import episodes_dao
 from . import *
 
-def create_or_update_listening_history(episode_id, user):
-  optional_episode = episodes_dao.get_episode(episode_id, user.id)
-  if optional_episode:
-    optional_listening_history = ListeningHistory.\
-      query.\
-      filter(
-          ListeningHistory.episode_id == episode_id,
-          ListeningHistory.user_id == user.id
-      ).first()
-
-    if optional_listening_history:
-      optional_listening_history.update_listening_time()
-      db_utils.db_session_commit()
-      return optional_listening_history
-
-    listening_history = ListeningHistory(episode_id=episode_id, user_id=user.id)
-    return db_utils.commit_model(listening_history)
-  else:
-    raise Exception('Invalid episode_id provided')
+def create_or_update_listening_histories(episode_update_info_map, user):
+  episode_ids = episode_update_info_map.keys()
+
+  listening_histories = \
+    ListeningHistory.query.filter(ListeningHistory.episode_id.in_(episode_ids),
+                                  ListeningHistory.user_id == user.id).all()
+  listening_histories_by_episode = \
+    {lh.episode_id: lh for lh in listening_histories}
+  result = []
+  for episode_id in episode_ids:
+    try:
+      current_progress = \
+        episode_update_info_map[episode_id]['current_progress']
+      percentage_listened = \
+        episode_update_info_map[episode_id]['percentage_listened']
+    except KeyError:
+      raise Exception('Invalid episode_id {} provided'.format(episode_id))
+    if episode_id in listening_histories_by_episode:
+      lh = listening_histories_by_episode[episode_id]
+      lh.current_progress = current_progress
+      lh.percentage_listened = lh.percentage_listened + percentage_listened
+    else:
+      lh = \
+        ListeningHistory(episode_id=episode_id, user_id=user.id,
+                         percentage_listened=percentage_listened,
+                         current_progress=current_progress)
+      db.session.add(lh)
+    result.append(lh)
+  # Update episode durations if needed
+  episode_ids_to_query = []
+  for episode_id, values in episode_update_info_map.iteritems():
```

Comment:
Maybe try re-writing this with zip() and by filtering on keys. Use some functional programming concepts here instead of this savage array work here, checking on values.

Reply:
Nice point, this can be compressed to a pretty simple `[k for k, v in episode_update_info_map.items() if 'real_duration' in v]`